### ISSUE TITLE
`slice::Iter` regression since Rust 1.65 (LLVM 15)

### ISSUE THREAD

**Issue by seritools** [rust-lang/rust#129583]
### Code

I tried this code:

```rust
pub fn fna(x: [u8; 1024]) -> u8 {
    x.into_iter().max().unwrap()
}

pub fn fnb(x: [u8; 1024]) -> u8 {
    *x.iter().max().unwrap()
}
```

[Godbolt](https://rust.godbolt.org/z/f36snahTn)

I expected to see this happen:

1. Both functions should optimize away the unwrap
2. (Bonus) Both functions should inline the respective iterator functions and auto-vectorize

Instead, this happened:

1. `fna` has no unwrap in the assembly output, but `fnb` has.
2. (Bonus) Both inline their respective iterators (according to the MIR output on the playground). `fna` auto-vectorizes, but `fnb` doesn't.

`fnb` generates a byte-wise max implementation. Interestingly, it keeps _some_ information about the array length around, influencing the codegen. A byte-wise `max` implementation for `[u8; 1024]` will need to do 1023 comparisons (since the first value of the array will be the initial `max` result value), and 1023 is divisible by 3, so the compiler emits a loop that emits a loop that runs 341 times, doing 3 `max` computations each. If I change the length of the array to 1025, it will do 4 `max` computations each.

Even though the generated code relies on the number of array entries, it still doesn't elite the `Option::unwrap` check for the `max` result. In Rust 1.45-1.64 `fnb` also generates the 3-byte-at-a-time loop, but _does_ remove the `Option::unwrap`.

### Version it worked on

It most recently worked on: Rust 1.64. (For context: Rust 1.65 added LLVM 15.)

It (initially?) started working in 1.45 (that version added LLVM 10).

### Version with regression

Any version since Rust 1.65.

-----

Since I thought this behavior was unexpected, I polled it on mastodon before looking further into it. ATM about half of the respondents thought just like me that both should be optimized (at least the unwrap part):

![image](https://github.com/user-attachments/assets/809ff640-b7b2-42dd-98af-a28aaa1706f2)

**Comment by the8472** [comment]
`fnb` is a by-reference iterator. If you use `.copied().max()` it vectorizes. See #106539

**Comment by seritools** [comment]
@the8472 ah, that explains the vectorization difference (bonus). just to clarify - that one is not the regression, just another thing I've noticed ^^

**Comment by saethlin** [comment]
The slice by-reference iterator fully unrolls and optimizes out the unwrap for `len <= 10`: https://rust.godbolt.org/z/1cT4z91eP

That's not a power of 2 so I feel like the distinction between 10 and 11 is a magic number in some LLVM optimization.

**Comment by saethlin** [comment]
It looks like this is a difference in InstCombine: https://rust.godbolt.org/z/hKof6449W

**Comment by nikic** [comment]
@saethlin Yeah, the problem there is that we need to look through a very deep sequence of selects to determine that the final pointer is actually guaranteed to be non-null. This is going to hit a cutoff sooner or later.

I think we could support this without cutoffs inside SCCP, by starting from a NotConstant(Null) value lattice root for the nonnull argument and then propagating that across GEP inbounds. I'll try that when I have a chance.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23129583.20.60slice.3A.3AIter.60.20regression.20since.20Rust.201.2E65.20.28LLVM.2015.29)).

@rustbot label -I-prioritize +P-low

**Comment by nikic** [comment]
https://github.com/llvm/llvm-project/pull/106090 should fix both the unrolled and not unrolled cases.
